File: frontend/src/services/authService.ts
import { supabase } from '@/lib/supabase';
import { AuthResponse, LoginRequest, RegisterRequest, ChangePasswordRequest, User } from '@/types/auth';

export class AuthService {
  static async login(credentials: LoginRequest): Promise<AuthResponse> {
    const { data, error } = await supabase.auth.signInWithPassword({
      email: credentials.email,
      password: credentials.password,
    });

    if (error) throw error;
    if (!data.user || !data.session) throw new Error('No user data returned');

    return this.getUserProfile(data.user, data.session.access_token);
  }

  static async register(data: RegisterRequest): Promise<AuthResponse> {
    const { data: authData, error } = await supabase.auth.signUp({
      email: data.email,
      password: data.password,
      options: {
        data: {
          full_name: data.full_name,
        },
      },
    });

    if (error) throw error;
    if (!authData.user) throw new Error('Registration successful but no user data returned');

    // If email confirmation is enabled, session might be null
    const token = authData.session?.access_token || '';

    return this.getUserProfile(authData.user, token);
  }

  static async logout(): Promise<void> {
    const { error } = await supabase.auth.signOut();
    if (error) throw error;
    localStorage.removeItem('auth_token');
    localStorage.removeItem('user_data');
  }

  static async getCurrentUser(): Promise<User> {
    const { data: { user }, error } = await supabase.auth.getUser();

    if (error) throw error;
    if (!user) throw new Error('No user logged in');

    const authResponse = await this.getUserProfile(user, '');
    return authResponse; // AuthResponse includes User fields
  }

  static async updateProfile(data: Partial<User>): Promise<User> {
    const updates: any = {};
    if (data.full_name) updates.data = { full_name: data.full_name };
    if (data.email) updates.email = data.email;

    const { data: { user }, error } = await supabase.auth.updateUser(updates);

    if (error) throw error;
    if (!user) throw new Error('Update failed');

    // Also update public.users if needed via API or Supabase client
    if (Object.keys(data).length > 0) {
      const { error: profileError } = await supabase
        .from('users')
        .update({
          full_name: data.full_name,
          // Add other fields as needed
        })
        .eq('id', user.id);

      if (profileError) console.error("Failed to update public profile:", profileError);
    }

    const authResponse = await this.getUserProfile(user, '');
    return authResponse;
  }

  static async changePassword(data: ChangePasswordRequest): Promise<void> {
    const { error } = await supabase.auth.updateUser({
      password: data.new_password
    });

    if (error) throw error;
  }

  static getStoredToken(): string | null {
    // Check auth_token first (kept in sync by api.ts)
    const authToken = localStorage.getItem('auth_token');
    if (authToken && authToken !== 'null' && authToken !== 'undefined') {
      return authToken;
    }
    // Check Supabase internal key patterns
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key && key.startsWith('sb-') && key.endsWith('-auth-token')) {
        try {
          const data = JSON.parse(localStorage.getItem(key) || '');
          if (data?.access_token) return data.access_token;
        } catch { /* skip */ }
      }
    }
    return null;
  }

  static async getStoredTokenAsync(): Promise<string | null> {
    try {
      const { data } = await supabase.auth.getSession();
      if (data?.session?.access_token) {
        localStorage.setItem('auth_token', data.session.access_token);
        return data.session.access_token;
      }
    } catch (e) {
      console.warn('getStoredTokenAsync failed:', e);
    }
    return this.getStoredToken();
  }

  static getStoredUser(): AuthResponse | null {
    const userData = localStorage.getItem('user_data');
    return userData ? JSON.parse(userData) : null;
  }

  static isAuthenticated(): boolean {
    return !!this.getStoredToken();
  }

  // Helper to fetch full user profile including role from public table
  private static async getUserProfile(sbUser: any, token: string): Promise<AuthResponse> {
    // Fetch profile from public.users to get the real role
    const { data: profile, error } = await supabase
      .from('users')
      .select('*')
      .eq('id', sbUser.id)
      .single();

    if (error) {
      console.warn('Failed to fetch user profile, falling back to metadata:', error);
    }

    const user: User = {
      user_id: sbUser.id,
      email: sbUser.email || '',
      full_name: profile?.full_name || sbUser.user_metadata?.full_name || '',
      role: profile?.role || sbUser.user_metadata?.role || 'user',
      package: profile?.package || sbUser.user_metadata?.package || 'free',
      created_at: sbUser.created_at,
      is_active: true,
      auth_provider: 'supabase',
      email_verified: !!sbUser.email_confirmed_at,
      two_factor_enabled: false,
    };

    const response = {
      access_token: token,
      token_type: 'bearer',
      ...user
    };

    // Update local storage
    if (token) {
      localStorage.setItem('auth_token', token);
      localStorage.setItem('user_data', JSON.stringify(response));
    }

    return response;
  }
}

File: Zex/backend/services/auth_service.py
"""
Authentication Service
Handles user registration, login, and token management using Supabase
"""
from fastapi import HTTPException, status
from typing import Optional, Dict, Any
import logging
import asyncio

from core.database import get_db
from core.logger import app_logger as logger
from core.email_service import get_email_service

class AuthService:
    """Supabase authentication service"""
    
    @staticmethod
    async def register_user(email: str, password: str, full_name: str = "") -> Dict[str, Any]:
        """Register a new user using Supabase Auth"""
        db = await get_db()
        
        try:
            # Sign up with Supabase
            response = db.auth.sign_up({
                "email": email,
                "password": password,
                "options": {
                    "data": {
                        "full_name": full_name,
                        "role": "user",
                        "package": "free"
                    }
                }
            })
            
            if not response.user:
                raise Exception("Registration failed")
                
            user = response.user
            
            # Send welcome email (async, don't block registration)
            try:
                email_service = get_email_service()
                asyncio.create_task(
                    email_service.send_welcome_email(
                        user_email=email,
                        user_name=full_name or email.split('@')[0],
                        credits=100  # Default starting credits
                    )
                )
                logger.info(f"Welcome email queued for {email}")
            except Exception as email_error:
                logger.warning(f"Failed to send welcome email: {email_error}")
            
            # Return auth response structure
            return {
                "user_id": user.id,
                "email": user.email,
                "full_name": user.user_metadata.get("full_name", ""),
                "role": user.user_metadata.get("role", "user"),
                "package": user.user_metadata.get("package", "free"),
                "access_token": response.session.access_token if response.session else None,
                "token_type": "bearer"
            }
            
        except Exception as e:
            logger.error(f"Registration error: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=str(e)
            )
    
    @staticmethod
    async def login_user(email: str, password: str) -> Dict[str, Any]:
        """Login user using Supabase Auth"""
        db = await get_db()
        
        try:
            response = db.auth.sign_in_with_password({
                "email": email,
                "password": password
            })
            
            if not response.user or not response.session:
                raise Exception("Login failed")
                
            user = response.user
            
            return {
                "user_id": user.id,
                "email": user.email,
                "full_name": user.user_metadata.get("full_name", ""),
                "role": user.user_metadata.get("role", "user"),
                "package": user.user_metadata.get("package", "free"),
                "access_token": response.session.access_token,
                "token_type": "bearer"
            }
            
        except Exception as e:
            logger.error(f"Login error: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid email or password"
            )
    
    @staticmethod
    async def get_user_by_id(user_id: str) -> Optional[Dict[str, Any]]:
        """Get user by ID from Supabase"""
        db = await get_db()
        try:
            # Try to get from public.users table first (if it exists and is synced)
            response = db.table("users").select("*").eq("id", user_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception:
            return None
    
    @staticmethod
    async def change_password(user_id: str, new_password: str) -> bool:
        """Change user password using Supabase Admin"""
        db = await get_db()
        try:
            # Use admin API to update user password
            response = db.auth.admin.update_user_by_id(
                user_id,
                {"password": new_password}
            )
            return bool(response.user)
        except Exception as e:
            logger.error(f"Password change error: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to change password"
            )

    @staticmethod
    async def update_user_profile(user_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Update user profile in public.users table"""
        db = await get_db()
        try:
            # Update public.users table
            response = db.table("users").update(data).eq("id", user_id).execute()
            
            # If full_name is updated, also update auth metadata
            if "full_name" in data:
                db.auth.admin.update_user_by_id(
                    user_id,
                    {"user_metadata": {"full_name": data["full_name"]}}
                )
                
            if response.data:
                return response.data[0]
            raise Exception("Profile update returned no data")
            
        except Exception as e:
            logger.error(f"Profile update error: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to update profile"
            )
